syserol/tensor.py

```diff
@@ -181,18 +185,73 @@ def perform_CMTF(tOrig=None, mOrig=None, r=8):
         tFac.mFactor = np.linalg.lstsq(tFac.factors[0][selPat, :], mOrig[selPat, :], rcond=None)[0].T
 
         if ii % 2 == 0:
-            R2X_last = R2X
-            R2X = calcR2X(tFac, tOrig, mOrig)
+            R2X_last = tFac.R2X
+            tFac.R2X = calcR2X(tFac, tOrig, mOrig)
+            assert tFac.R2X > 0.0
 
-        if R2X - R2X_last < 1e-9:
+        if tFac.R2X - R2X_last < 1e-9:
             break
 
+    # Refine with direct optimization
+    tFac = fit_refine(tFac, tOrig, mOrig)
+
     tFac = cp_normalize(tFac)
     tFac = reorient_factors(tFac)
-    tFac.R2X = R2X
+    tFac = sort_factors(tFac)
 
     if pick:
         with open(filename, "wb") as p:
-            pickle.dump(sort_factors(tFac), p)
+            pickle.dump(tFac, p)
+
+    return tFac
+
+
+def cp_to_vec(tFac):
+    vec = np.concatenate([tFac.factors[i].flatten() for i in range(3)])
+    return np.concatenate((vec, tFac.mFactor.flatten()))
+
+
+def buildTensors(pIn, tensor, matrix, r):
+    """ Use parameter vector to build kruskal tensors. """
+    assert tensor.shape[0] == matrix.shape[0]
+    nA = tensor.shape[0]*r
+    nB = tensor.shape[1]*r
+    nC = tensor.shape[2]*r
+    A = np.reshape(pIn[:nA], (tensor.shape[0], r))
+    B = np.reshape(pIn[nA:nA+nB], (tensor.shape[1], r))
+    C = np.reshape(pIn[nA+nB:nA+nB+nC], (tensor.shape[2], r))
+    tFac = tl.cp_tensor.CPTensor((None, [A, B, C]))
+    tFac.mFactor = np.reshape(pIn[nA+nB+nC:], (matrix.shape[1], r))
+    tFac.mWeights = np.ones(r)
+    return tFac
 
-    return sort_factors(tFac)
+
+def cost(pIn, tOrig, mOrig, r):
+    tFac = buildTensors(pIn, tOrig, mOrig, r)
+    tDiff = np.nan_to_num(tOrig - tl.kruskal_to_tensor(tFac))
+    mDiff = np.nan_to_num(mOrig - buildGlycan(tFac))
+
+    # Overall cost
+    cost = 0.5 * (np.linalg.norm(tDiff) + np.linalg.norm(mDiff))
+
+    gtFac = deepcopy(tFac)
+    gtFac.factors = [-unfolding_dot_khatri_rao(tDiff, tFac, ii) for ii in range(3)]
+
+    mCP = (None, [tFac.factors[0], tFac.mFactor])
+    gtFac.factors[0] += -unfolding_dot_khatri_rao(mDiff, mCP, 0)
+    gtFac.mFactor = -unfolding_dot_khatri_rao(mDiff, mCP, 1)
+
+    return cost, cp_to_vec(gtFac)
+
+
+def fit_refine(tFac, tOrig, mOrig):
+    """ Refine the factorization with direct optimization. """
+    r = tFac.rank
+
+    x0 = cp_to_vec(tFac)
+    res = minimize(cost, x0, jac=True, args=(tOrig, mOrig, r), options={"disp": True, "maxiter": 3000})
```

Comment:
The reason I suspect the gradients are wrong is this exits after 0 iterations. I also compared my gradient to numerical differencing with [`scipy.optimize.approx_fprime`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.approx_fprime.html).